macaroonbakery/oven.py

```diff
@@ -0,0 +1,239 @@
+
+    def add_caveat(self, m, cav):
+        '''Add a caveat to the given macaroon.
+        '''
+        return m.add_caveat(cav, self.key, self.locator)
+
+    def add_caveats(self, m, cavs):
+        '''Add all the caveats to the given macaroon.
+        '''
+        return m.add_caveats(cavs, self.key, self.locator)
+
+    def macaroon_id(self, ops, storage_id, expiry):
+        nonce = os.urandom(16)
+        if len(ops) == 1 or self.ops_store is None:
+            return id_pb2.MacaroonId(
+                nonce=nonce,
+                storageId=storage_id,
+                ops=_macaroon_id_ops(ops))
+        # We've got several operations and a multi-op store, so use the store.
+        # TODO use the store only if the encoded macaroon id exceeds some size?
+        entity = self.ops_entity(ops)
+        self.ops_store.put_opts(entity, ops, expiry)
+        return id_pb2.MacaroonId(
+            nonce=nonce,
+            storageId=storage_id,
+            ops=[id_pb2.Op(entity=entity, actions=["*"])])
+
+    def ops_entity(self, ops):
+        ''' Returns a new multi-op entity name string that represents
+        all the given operations and caveats. It returns the same value
+        regardless of the ordering of the operations. It assumes that the
+        operations have been canonicalized and that there's at least one
+        operation.
+
+        :param ops:
+        :return: string that represents all the given operations and caveats.
+        '''
+        # Hash the operations, removing duplicates as we go.
+        hash_entity = hashlib.sha256()
+        for op in ops:
+            hash_entity.update('{}\n{}\n'.format(
+                op.action, op.entity).encode())
+        hash_encoded = base64.urlsafe_b64encode(hash_entity.digest())
+        return 'multi-' + hash_encoded.decode('utf-8').rstrip('=')
+
+    def macaroon_ops(self, macaroons):  # (ops []Op, conditions []string) {
+        '''
+
+        For macaroons minted with previous bakery versions, it always
+        returns a single LoginOp operation.
+
+        :param macaroons:
+        :return:
+        '''
+        if len(macaroons) == 0:
+            raise ValueError("no macaroons provided")
+
+        storage_id, ops = _decode_macaroon_id(macaroons[0].identifier_bytes)
+        root_key = self.root_keystore_for_ops(ops).get(storage_id)
+        if root_key is None:
+            raise ValueError("macaroon not found in storage")
+        v = Verifier()
+        conditions = []
+
+        def validator(condition):
+            conditions.append(condition)
+            return True
+        v.satisfy_general(validator)
+        v.verify(macaroons[0], root_key, macaroons[1:])
+        if (self.ops_store is not None
+            and len(ops) == 1
+                and ops[0].entity.startswith("multi-")):
+            # It's a multi-op entity, so retrieve the actual operations
+            # it's associated with.
+            ops = self.ops_store.get_opts(ops[0].entity)
```

Comment:
s/get_opts/get_ops/

Reply:
Maybe get_ops should raise an exception when the entity doesn't exist, rather than returning None?

Reply:
done.